tools/database/utils.go

```diff
@@ -62,3 +65,135 @@ func ParseRandomSuffixOptions(optionsString sql.NullString) ([]int32, error) {
+func formatStrings(maxLength int, input []string) []string {
+	result := []string{}
+	for _, line := range input {
+		words := strings.Split(strings.Trim(line, "\n\r "), " ")
+		currentLine := ""
+		for _, word := range words {
+			if len(currentLine) > maxLength {
+				result = append(result, currentLine)
+				currentLine = ""
+			}
+
+			if len(currentLine) > 0 {
+				currentLine += " "
+			}
+
+			currentLine += word
+		}
+
+		if len(result) > 0 || len(currentLine) > 0 {
+			result = append(result, currentLine)
+		}
+	}
+
+	return result
+}
+
+func parseDungeonDifficultyMask(mask int) proto.DungeonDifficulty {
+	// for negative masks take the 2 compliment value of the lower 8 bits
+	if mask < 0 {
+		mask = mask & 0xFF
+	}
+
+	// we only map to one difficulty so for now do the best match
+	if mask&dbc.LOOKING_FOR_RAID > 0 {
+		return proto.DungeonDifficulty_DifficultyRaid25RF
+	}
+
+	if mask&dbc.HEROIC_RAID_25_MAN > 0 {
+		return proto.DungeonDifficulty_DifficultyRaid25H
+	}
+
+	if mask&dbc.HEROIC_RAID_10_MAN > 0 {
+		return proto.DungeonDifficulty_DifficultyRaid10H
+	}
+
+	if mask&dbc.NORMAL_RAID_25_MAN > 0 {
+		return proto.DungeonDifficulty_DifficultyRaid25
+	}
+
+	if mask&dbc.NORMAL_RAID_10_MAN > 0 {
+		return proto.DungeonDifficulty_DifficultyRaid10
+	}
+
+	if mask&dbc.HEROIC_DUNGEON > 0 {
+		return proto.DungeonDifficulty_DifficultyHeroic
+	}
+
+	if mask&dbc.NORMAL_DUNGEON > 0 {
+		return proto.DungeonDifficulty_DifficultyNormal
+	}
+
+	return proto.DungeonDifficulty_DifficultyUnknown
+}
+
+func DifficultyToShortName(difficulty proto.DungeonDifficulty) string {
+	switch difficulty {
+	case proto.DungeonDifficulty_DifficultyHeroic, proto.DungeonDifficulty_DifficultyRaid10H, proto.DungeonDifficulty_DifficultyRaid25H:
+		return "(H)"
+	case proto.DungeonDifficulty_DifficultyNormal, proto.DungeonDifficulty_DifficultyRaid25, proto.DungeonDifficulty_DifficultyRaid10:
+		return "(N)"
+	case proto.DungeonDifficulty_DifficultyRaid25RF:
+		return "(LFR)"
+	default:
+		return ""
+	}
+}
+
+func EnchantHasDummyEffect(enchant *proto.UIEnchant, instance *dbc.DBC) bool {
+	return SpellHasDummyEffect(int(enchant.SpellId), instance)
+}
+
+func SpellHasDummyEffect(spellId int, instance *dbc.DBC) bool {
+	if effects, ok := instance.SpellEffects[spellId]; ok {
+		for _, effect := range effects {
+			if effect.EffectAura == dbc.A_DUMMY ||
+				effect.EffectAura == dbc.A_PERIODIC_DUMMY {
+				return true
+			}
+		}
+	}
+
+	return false
+}
+
+func SpellHasTriggerEffect(spellId int, instance *dbc.DBC) bool {
+	if effects, ok := instance.SpellEffects[spellId]; ok {
+		for _, effect := range effects {
+			if effect.EffectAura == dbc.A_PROC_TRIGGER_SPELL ||
+				effect.EffectAura == dbc.A_PROC_TRIGGER_SPELL_WITH_VALUE {
+				return true
+			}
+		}
+	}
+
+	return false
+}
+
+func SpellUsesStacks(spellId int, instance *dbc.DBC) bool {
+	if spell, ok := instance.Spells[spellId]; ok {
+		if spell.MaxCumulativeStacks > 1 {
+			return true
+		}
```

Comment:
Is the stat bonus per stack also contained in the DB? If so, could we generalize `shared.ProcStatBonusEffect` to handle both stacking and non-stacking buffs?

Reply:
Ah okay, if it's a mix then it might be better to stick with always defining those ones manually. But in that case we'll definitely need to have `shared.StackingStatBonusEffect` support scaling items.